packages/graphback-core/src/relationships/RelationshipMetadataBuilder.ts

```diff
@@ -0,0 +1,334 @@
+        }
+
+        const relationFieldBaseType = getBaseType(relationField.type);
+
+        if (!isObjectType(relationFieldBaseType) || relationFieldBaseType.name !== modelName) {
+            throw new Error(`${modelName}.${field.name} relationship field maps to ${relationType.name}.${relationField.name} (${relationFieldBaseType.name} type) which should be ${modelName} type.`);
+        }
+
+        const manyToOneMetadata = parseRelationshipAnnotation(relationField.description);
+
+        if (manyToOneMetadata && oneToManyMetadata.key && manyToOneMetadata.key && oneToManyMetadata.key !== manyToOneMetadata.key) {
+            throw new Error(`${modelName}.${field.name} and ${relationType.name}.${relationField.name} 'key' annotations are different. Ensure both are the same, or remove one so that it can be generated.`);
+        }
+
+        // this.validateManyToOneField(relationType.name, relationField);
+    }
+
+    private validateManyToOneField(modelName: string, field: GraphQLField<any, any>) {
+        this.validateField(modelName, field);
+
+        const metadata = parseRelationshipAnnotation(field.description);
+
+        if (metadata.kind !== 'manyToOne') {
+            throw new Error(`${modelName}.${field.name} should be a @manyToOne field, but has a @${metadata.kind} annotation`);
+        }
+    }
+
+    private validateOneToOneRelationship(modelName: string, field: GraphQLField<any, any>) {
+        this.validateField(modelName, field);
+
+        const metadata = parseRelationshipAnnotation(field.description);
+
+        if (metadata.kind !== 'oneToOne') {
+            throw new Error(`${modelName}.${field.name} should be a @oneToOne field, but has a ${metadata.kind} annotation`);
+        }
+
+        const relationType = getBaseType(field.type) as GraphQLObjectType;
+        const relationField = relationType.getFields()[metadata.field];
+
+        // field will be generated
+        if (!relationField) {
+            return;
+        }
+
+        if (hasListType(relationField.type)) {
+            throw new Error(`${relationType.name}.${relationField.name} is a list type, but should be '${relationField.name}: ${modelName}'.`)
+        }
+
+        const relationFieldBaseType = getBaseType(relationField.type);
+
+        if (!isObjectType(relationFieldBaseType) || relationFieldBaseType.name !== modelName) {
+            throw new Error(`${modelName}.${field.name} relationship field maps to ${relationType.name}.${relationField.name} (${relationFieldBaseType.name} type) which should be ${modelName} object type.`);
+        }
+    }
+
+    private validateField(modelName: string, field: GraphQLField<any, any>) {
+        const relationshipAnnotation = parseRelationshipAnnotation(field.description);
+        if (!relationshipAnnotation) {
+            throw new Error(`${modelName}.${field.name}`)
```

Comment:
So I guess we made a call to not build validation into the engine and each plugin will simply respond to throw errors - this will mean that we meed to close created issue and add some note to the docs

Reply:
Yeah I thought about it and this way seemed better to me. Separating validation from execution means scanning the data model twice. It also means separating closely related operations.

It would be very hard to implement relationship validation into the engine because a user could swap the SchemaCRUDPlugin for their own that uses directives. This makes the pre-validation...invalid.